src/core/internal/hash.d

```diff
@@ -1010,6 +1010,41 @@ private template bytesHashAlignedBy(AlignType)
     alias bytesHashAlignedBy = bytesHash!(AlignType.alignof >= uint.alignof);
 }
 
+private template bytesHashWithExactSizeAndAlignment(SizeAndAlignType)
+{
+    // Don't take into account alignment in the choice between bytesHash
+    // and smallBytesHash so the logic in TypeInfo_struct will not be
+    // complicated. Taking it into account the threshold with uint-aligned
+    // data would be ~10 bytes and the threshold with non-aligned data
+    // would be ~12 bytes.
+    static if (SizeAndAlignType.sizeof <= smallBytesHashThreshold)
+        alias bytesHashWithExactSizeAndAlignment = smallBytesHash;
+    else
+        alias bytesHashWithExactSizeAndAlignment = bytesHashAlignedBy!SizeAndAlignType;
+}
+
+// Size at & below which to prefer `smallBytesHash` over `bytesHash`.
+// Internal. Not to be used other than by druntime.
+enum smallBytesHashThreshold = 10;
+
+// Internal. Not to be used other than by druntime.
+@nogc nothrow pure @trusted
+size_t smallBytesHash()(scope const(ubyte)[] bytes, size_t seed)
```

Comment:
Does this and the above symbols need to be public?

Reply:
I would prefer not, but I needed to make it public to call it from `object.TypeInfo_struct`.